text/0000-build-std/4-stage-1a.md

```diff
@@ -0,0 +1,1010 @@
+[rationale-incremental]: #why-not-build-the-standard-library-in-incremental
+
+As the standard library sources are never modified, incremental compilation
+would only add a compilation time overhead.
+
+↩ [*Proposal*][proposal]
+
+### Why not produce a `dylib` for the standard library?
+[rationale-no-dylib]: #why-not-produce-a-dylib-for-the-standard-library
+
+The `std` crate's `Cargo.toml` is configured with
+`crate-type = ["rlib", "dylib"]` so it can produce both artifacts. The Rust
+project ships both artifacts, with the `dylib` only linked against when
+`-Cprefer-dynamic` is enabled. However, the `dylib` is not part of Rust's
+stability guarantee so a first-class way of specifying crate types is left to a
+future extension.
+
+*See the following sections for future possibilities:*
+
+- [*Allow choosing the crate type of the standard library?*][future-crate-type]
+
+↩ [*Proposal*][proposal]
+
+## Why use the pre-built standard library for procedural macros and build-scripts?
+[rationale-sysroot-for-host-deps]: #why-use-the-pre-built-standard-library-for-procedural-macros-and-build-scripts
+
+Procedural macros and build scripts always run on the host and need to be built
+with a configuration that are compatible with the host toolchain's Cargo and
+rustc. There is little advantage to using a custom standard library with
+procedural macros or build scripts, as they are not part of the final output
+artifact and anywhere they can run already have a toolchain with host tools and
+a pre-built standard library. Procedural macros must link against the compiler
+which further limits potential use cases to those without `target-modifiers`.
+
+↩ [*Proposal*][proposal]
+
+## Why not replace `#![no_std]` as the source-of-truth for whether a crate depends on `std`?
+[rationale-replace-no_std]: #why-not-replace-no_std-as-the-source-of-truth-for-whether-a-crate-depends-on-std
+
+Crates can currently use the crate attribute `#![no_std]` to indicate a lack of
+dependency on `std`. With `Cargo.toml` being used to express a dependency on the
+standard library (or lack thereof), it is unintuitive for there to be two
+sources-of-truth for this information.
+
+`#![no_std]` serves two purposes - it stops the compiler from loading `std` from
+the sysroot and adding `extern crate std`, and it prevents the user from
+depending on anything from `std` accidentally.
+
+`#![no_std]` could hypothetically be replaced by a lint to prevent use of the
+standard library and a change to the compiler so that it loads the `std`
+speculatively unless it is used.
+
+However, while rustc does have some support for speculatively loading crates, it
+is not possible to do so and not declare them as a dependency in cross-crate
+metadata.
+
+↩ [*Interactions with `#![no_std]`*][interactions-with-no_std]
+
+### Why remove `restricted_std`?
+[rationale-remove-restricted-std]: #why-remove-restricted_std
+
+`restricted_std` was originally added as part of a mechanism to enable the
```

Comment:
restricted_std was added to gate the standard library on certain targets, which allows the standard library to somewhat support more targets with stubbed out functionality. This gating behaviour still needs to be there. The old `--print std-support` we had worked fine, but it would be ideal if something did this gating for non-cargo users.